napari/layers/_shapes_layer/shapes/shape.py

```diff
@@ -0,0 +1,297 @@
+        self.edge_color = edge_color
+        self.face_color = face_color
+        self.opacity = opacity
+        self.z_index = z_index
+        self.name = ''
+
+    @property
+    @abstractmethod
+    def data(self):
+        # user writes own docstring
+        raise NotImplementedError()
+
+    @data.setter
+    @abstractmethod
+    def data(self, data):
+        raise NotImplementedError()
+
+    @property
+    def edge_width(self):
+        """float: thickness of lines and edges.
+        """
+        return self._edge_width
+
+    @edge_width.setter
+    def edge_width(self, edge_width):
+        self._edge_width = edge_width
+
+    @property
+    def edge_color(self):
+        """Color, ColorArray: color of edges
+        """
+        return self._edge_color
+
+    @edge_color.setter
+    def edge_color(self, edge_color):
+        self._edge_color = Color(edge_color)
+        if type(edge_color) is str:
+            self._edge_color_name = edge_color
+        else:
+            rgb = tuple([int(255*x) for x in self._edge_color.rgba[:3]])
+            self._edge_color_name = '#%02x%02x%02x' % rgb
+
+    @property
+    def face_color(self):
+        """Color, ColorArray: color of faces
+        """
+        return self._face_color
+
+    @face_color.setter
+    def face_color(self, face_color):
+        self._face_color = Color(face_color)
+        if type(face_color) is str:
+            self._face_color_name = face_color
+        else:
+            rgb = tuple([int(255*x) for x in self._face_color.rgba[:3]])
+            self._face_color_name = '#%02x%02x%02x' % rgb
+
+    @property
+    def opacity(self):
+        """float: opacity of shape
+        """
+        return self._opacity
+
+    @opacity.setter
+    def opacity(self, opacity):
+        self._opacity = opacity
+
+    @property
+    def z_index(self):
+        """int: z order priority of shape. Shapes with higher z order displayed
+        ontop of others.
+        """
+        return self._z_index
+
+    @z_index.setter
+    def z_index(self, z_index):
+        self._z_index = z_index
+
+    def _set_meshes(self, data, closed=True, face=True, edge=True):
+        """Sets the face and edge meshes from a set of points.
+
+        Parameters
+        ----------
+        data : np.ndarray
+            Nx2 array specifying the shape to be triangulated
+        closed : bool
+            Bool which determines if the edge is closed or not
+        face : bool
+            Bool which determines if the face need to be traingulated
+        edge : bool
+            Bool which determines if the edge need to be traingulated
+        """
+        if edge:
+            centers, offsets, triangles = triangulate_edge(data, closed=closed)
+            self._edge_vertices = centers
+            self._edge_offsets = offsets
+            self._edge_triangles = triangles
+        if face:
+            if len(data) > 2:
+                vertices, triangles = triangulate_face(data)
```

Comment:
I've seen this pattern pop up in a few places. It could be worth moving it into `shape_util`

```python
if len(data) > 2:
                vertices, triangles = triangulate_face(data)
                
                 if len(triangles) > 0:
                    self._face_vertices = vertices
                    self._face_triangles = triangles
```

Reply:
hmm, not sure about this - searching for `triangulate_face` only finds it used here